compiler/rustc_middle/src/ty/trait_def.rs

```diff
@@ -66,11 +64,11 @@ pub enum TraitSpecializationKind {
     AlwaysApplicable,
 }
 
-#[derive(Default, Debug)]
+#[derive(Default, Debug, HashStable)]
 pub struct TraitImpls {
     blanket_impls: Vec<DefId>,
     /// Impls indexed by their simplified self type, for fast lookup.
-    non_blanket_impls: FxHashMap<fast_reject::SimplifiedType, Vec<DefId>>,
+    non_blanket_impls: FxIndexMap<fast_reject::SimplifiedType, Vec<DefId>>,
```

Comment:
Hm, is this enough?

It looks like insertion into non_blanket_impls happens in a loop which is a result of `tcx.crates(())` which isn't stable across compilation sessions I'd expect (though I could be wrong).

Reply:
The way `crates(())` is computed, it looks like it is always `(1..=num_crates).map(CrateNum)` (0 is reserved for the current crate and does not have any data) and it I would assume the crates get added to the `CStore` in a deterministic order.

We do have `stable_crate_id` which orders crates independently of their `cnum`. I could be that `CrateNum`s are not deterministically assigned to their actual crate data, though it seems quite buggy to me if this doesn't invalidate query caches in that case. e.g. `trait_impls_of_provider` just uses the order of `crates(())` which influences the way winnowing works for example.

A bit out of my depth here, maybe @rust-lang/wg-incr-comp can help and clarify here :sweat_smile: :heart:

Reply:
Yeah, we probably do end up with an ordering that is mostly stable from run to run (presuming no shifts in code or other sources), and I think incremental will track the query fine. My guess is that we could avoid marking the trait query as red if the crate load order changes (e.g., because the first code using that crate is moved?) if this used a stable crate ID, but it may not be worth it. Ultimately this PR does seem like it makes things strictly better here :)

Reply:
I *think* that crate ordering is indeed deterministic, since it isn't something that we discover dynamically? But indeed I bet @michaelwoerister or @petrochenkov would be certain.

Reply:
Crate are assigned their CrateNum in the order in which they are required by the resolver. As the resolver is not incremental, this order is deterministic but not stable. On the other hand, it is not obvious at all.
Incremental computes a hash for `crates(())` based on their order in the vector and their `stable_crate_id`. We could make it independent on the order of crates by having the `crates` query sort its result by increasing `stable_crate_id`.

Reply:
I think that parts of the code might rely on `tcx.crates(())` being sorted by `CrateNum`, for better or worse. I would not recommend trying to change that as part of this PR.